File: src/physics/dynamics.py
"""
dynamics - Função de Dinâmica (para bomba "dumb" com modelo baseado em alpha e beta)
Descrição: Calcula as derivadas dos estados de uma bomba sem controle e sem propulsão,
           mas mantendo o modelo V, alpha, beta, etc.
"""

import numpy as np

# Importações necessárias (fora da função):
from models.Cmat import Cmat
from models.skew import skew  # <-- IMPORTAÇÃO DA FUNÇÃO SKEW AQUI
from physics.aero_loads import aero_loads
from physics.prop_loads import prop_loads
from physics.ISA import ISA
from globals import aircraft, g

def dynamics(t, X, U, W):
    """
    Calcula a dinâmica do sistema (bomba "dumb") mantendo V, alpha, beta, etc.

    Entrada:
    - t (float): tempo atual [s].
    - X (numpy.ndarray): vetor de estados (12 elementos) => [V, alpha, q, theta, H, x, beta, phi, p, r, psi, y]
    - U (numpy.ndarray): vetor de controles (6 elementos) [zerado ou ignorado neste caso].
    - W (numpy.ndarray): vetor de vento (3 elementos) [se não usar, passe zeros].

    Saída:
    - Xdot (numpy.ndarray): derivadas dos estados (12 elementos).
    - Y (numpy.ndarray): variáveis de saída para pós-processamento.
    """

    # ---------------------------------------------------------
    # 1) Extraindo estados
    # ---------------------------------------------------------
    V        = X[0]
    alpha_deg= X[1]
    q_deg_s  = X[2]
    theta_deg= X[3]
    H_m      = X[4]
    x        = X[5]
    beta_deg = X[6]
    phi_deg  = X[7]
    p_deg_s  = X[8]
    r_deg_s  = X[9]
    psi_deg  = X[10]
    y        = X[11]

    # Converte ângulos para radianos
    alpha_rad = np.deg2rad(alpha_deg)
    beta_rad  = np.deg2rad(beta_deg)
    phi_rad   = np.deg2rad(phi_deg)
    theta_rad = np.deg2rad(theta_deg)
    psi_rad   = np.deg2rad(psi_deg)

    p_rad_s = np.deg2rad(p_deg_s)
    q_rad_s = np.deg2rad(q_deg_s)
    r_rad_s = np.deg2rad(r_deg_s)

    # ---------------------------------------------------------
    # 2) Cálculo de (u, v, w) no referencial corpo
    # ---------------------------------------------------------
    u = V * np.cos(beta_rad) * np.cos(alpha_rad)
    v = V * np.sin(beta_rad)
    w = V * np.cos(beta_rad) * np.sin(alpha_rad)
    V_b = np.array([u, v, w])  # Velocidade no corpo

    # ---------------------------------------------------------
    # 3) Matrizes de rotação e gravidade
    # ---------------------------------------------------------
    C_phi   = Cmat(1, phi_rad)
    C_theta = Cmat(2, theta_rad)
    C_psi   = Cmat(3, psi_rad)
    # Transformação do veículo para o corpo:
    C_bv    = C_phi @ C_theta @ C_psi

    # Gravidade no referencial corpo
    g_b = C_bv @ np.array([0, 0, g])

    # ---------------------------------------------------------
    # 4) Propriedades físicas (bomba) e matrizes de massa
    # ---------------------------------------------------------
    m     = aircraft['m']
    J_O_b = aircraft['J_O_b']
    rC_b  = aircraft['rC_b']  # vetorzinho de CG

    Mgen = np.block([
        [m * np.eye(3),          -m * skew(rC_b)],
        [m * skew(rC_b),         J_O_b         ]
    ])

    # ---------------------------------------------------------
    # 5) Cargas aerodinâmicas e propulsivas (zeradas)
    # ---------------------------------------------------------
    # Se 'aero_loads' foi simplificada para a bomba, ela deve
    # retornar principalmente arrasto, e possivelmente sustentação pequena.
    Faero_b, Maero_O_b, Yaero = aero_loads(X, U)

    # Zerar a parte de propulsão (sem motor):
    Fprop_b = np.zeros(3)
    Mprop_O_b = np.zeros(3)

    # ---------------------------------------------------------
    # 6) Forças e momentos resultantes
    # ---------------------------------------------------------
    omega_b = np.array([p_rad_s, q_rad_s, r_rad_s])

    eq_F = m * skew(omega_b) @ V_b - m * skew(omega_b) @ skew(rC_b) @ omega_b
    eq_F += Faero_b + Fprop_b + m * g_b  # Fprop_b = 0

    eq_M = skew(omega_b) @ J_O_b @ omega_b
    eq_M += m * skew(rC_b) @ skew(omega_b) @ V_b
    eq_M += Maero_O_b + Mprop_O_b + m * skew(rC_b) @ g_b  # Mprop_O_b=0

    # Resolve o sistema [u_dot, v_dot, w_dot, p_dot, q_dot, r_dot]
    edot = np.linalg.solve(Mgen, np.concatenate([eq_F, eq_M]))
    u_dot, v_dot, w_dot = edot[:3]
    p_dot, q_dot, r_dot = edot[3:6]

    # ---------------------------------------------------------
    # 7) Cinemática angular e translacional
    # ---------------------------------------------------------
    # 7a) Velocidades angulares -> Euler angles
    HPhi_inv = np.column_stack((C_phi[:, 0], C_phi[:, 1], C_bv[:, 2]))
    Phi_dot_rad_s = np.linalg.solve(HPhi_inv, omega_b)

    # 7b) Translação no Eixo Inercial
    dREOdt = C_bv.T @ V_b  # dx/dt, dy/dt, dH/dt

    # ---------------------------------------------------------
    # 8) Derivadas de V, alpha, beta, etc.
    # ---------------------------------------------------------
    # V_dot
    if V > 1e-6:
        V_dot = (V_b @ edot[:3]) / V
    else:
        V_dot = 0.0

    # alpha_dot (em deg/s)
    denom_uw = u**2 + w**2
    if denom_uw > 1e-6:
        alpha_dot_rad_s = (u * w_dot - w * u_dot) / denom_uw
    else:
        alpha_dot_rad_s = 0.0
    alpha_dot_deg_s = np.rad2deg(alpha_dot_rad_s)

    # beta_dot (em deg/s)
    denom_uvw = V * np.sqrt(u**2 + w**2)
    if denom_uvw > 1e-6:
        beta_dot_rad_s = (V * v_dot - v * V_dot) / denom_uvw
    else:
        beta_dot_rad_s = 0.0
    beta_dot_deg_s = np.rad2deg(beta_dot_rad_s)

    # q_dot (em deg/s)
    q_dot_deg_s = np.rad2deg(q_dot)

    # p_dot, r_dot
    p_dot_deg_s = np.rad2deg(p_dot)
    r_dot_deg_s = np.rad2deg(r_dot)

    # theta_dot, phi_dot, psi_dot
    theta_dot_deg = np.rad2deg(Phi_dot_rad_s[1])
    phi_dot_deg   = np.rad2deg(Phi_dot_rad_s[0])
    psi_dot_deg   = np.rad2deg(Phi_dot_rad_s[2])

    # dH, dx, dy
    H_dot = dREOdt[2]
    x_dot = dREOdt[0]
    y_dot = dREOdt[1]

    # ---------------------------------------------------------
    # 9) Monta Xdot
    # ---------------------------------------------------------
    Xdot = np.array([
        V_dot,            # dV/dt
        alpha_dot_deg_s,  # d(alpha)/dt
        q_dot_deg_s,      # dq/dt
        theta_dot_deg,    # d(theta)/dt
        H_dot,            # dH/dt
        x_dot,            # dx/dt
        beta_dot_deg_s,   # d(beta)/dt
        phi_dot_deg,      # d(phi)/dt
        p_dot_deg_s,      # dp/dt
        r_dot_deg_s,      # dr/dt
        psi_dot_deg,      # d(psi)/dt
        y_dot             # dy/dt
    ])

    # ---------------------------------------------------------
    # 10) Variáveis de saída Y
    # ---------------------------------------------------------
    rho, _, _, a = ISA(H_m)
    Mach = V / a if a > 1e-6 else 0.0
    qbar = 0.5 * rho * (V**2)

    n_C_b = -1/(m*g) * (Faero_b + Fprop_b)  # Fprop_b=0
    r_pilot_b = aircraft['r_pilot_b']

    # Aceleração no CG do piloto (se quiser remover, fique à vontade)
    n_pilot_b = n_C_b - (1/g)*(
        skew(edot[3:6]) @ (r_pilot_b - rC_b) +
        skew(omega_b) @ skew(omega_b) @ (r_pilot_b - rC_b)
    )

    Y = np.concatenate([
        [V, alpha_deg, q_deg_s, theta_deg, H_m, x, beta_deg, phi_deg,
         p_deg_s, r_deg_s, psi_deg, y],
        n_pilot_b.flatten(),
        n_C_b.flatten(),
        [Mach, qbar],
        Fprop_b.flatten(),    # = 0
        Mprop_O_b.flatten(),  # = 0
        Yaero.flatten(),      # arrasto, sustentação, etc.
        [V_dot, alpha_dot_deg_s, beta_dot_deg_s, u, v, w]
    ])

    return Xdot, Y

File: docs/AB/Exame_CTR_Final/sim_nao_linear/dynamics.m
function [Xdot,Y] = dynamics(t,X,U)

global g
global aircraft
global rad2deg
% X = [u v w p q r x y z phi theta psi | P].';

u = X(1);
v = X(2);
w = X(3);
p_rad_s = X(4);
q_rad_s = X(5);
r_rad_s = X(6);
x = X(7);
y = X(8);
z = X(9);
phi_rad = X(10);
theta_rad = X(11);
psi_rad = X(12); 


omega_b = [p_rad_s q_rad_s r_rad_s].';
V_b = [u v w].';

C_phi = Cmat(1,phi_rad);
C_theta = Cmat(2,theta_rad);
C_psi = Cmat(3,psi_rad);

C_bv = C_phi*C_theta*C_psi;

g_b = C_bv*[0 0 g].';

m = aircraft.m;
J_O_b = aircraft.J_O_b;
rC_b = aircraft.rC_b;

Mgen = [m*eye(3) -m*skew(rC_b)
        m*skew(rC_b) J_O_b];

% Cargas aerodinâmicas e propulsivas:

[Faero_b,Maero_O_b,Yaero] = aero_loads(X,U);
[Fprop_b,Mprop_O_b,Yprop] = prop_loads(X,U);

% Termos restantes das EdMs:

% Primeiro membro (equação das forças):
eq_F = m*skew(omega_b)*V_b +...
    -m*skew(omega_b)*skew(rC_b)*omega_b;

% Segundo membro menos o primeiro:
eq_F = -eq_F +...
    Faero_b + Fprop_b + m*g_b;
                                
% Primeiro membro (equação dos momentos):

eq_M = skew(omega_b)*J_O_b*omega_b +...
    m*skew(rC_b)*skew(omega_b)*V_b;

% Segundo membro menos o primeiro:
eq_M = -eq_M +...
    Maero_O_b + Mprop_O_b + m*skew(rC_b)*g_b;

% Mgen*edot = [eq_F; eq_M];
edot = Mgen\[eq_F; eq_M];

% edot = [udot vdot wdot pdot qdot rdot]

% Cinemática de rotação:
HPhi_inv = [C_phi(:,1) C_phi(:,2) C_bv(:,3)];
Phi_dot_rad_s = HPhi_inv\omega_b; %Taxas angulares em rad/seg.

% Cinemática de translação:

dREOdt = C_bv.'*V_b;


% X = [u v w p q r x y z phi theta psi | P].';


V = Yaero(1);
alpha_deg = Yaero(2);
q_deg_s = q_rad_s*rad2deg;
theta_deg = theta_rad*rad2deg;
H = -z;

beta_deg = Yaero(3);
phi_deg = phi_rad*rad2deg;
p_deg_s = p_rad_s*rad2deg;
r_deg_s = r_rad_s*rad2deg;
psi_deg = psi_rad*rad2deg;

Vdot = (V_b.'*edot(1:3))/V;
udot = edot(1);
vdot = edot(2);
wdot = edot(3);
alpha_dot_rad_s = (u*wdot-w*udot)/(u^2+w^2);
beta_dot_rad_s = (V*vdot-v*Vdot)/(V*sqrt(u^2+w^2));

n_C_b = -1/(m*g)*(Faero_b + Fprop_b);

r_pilot_b = aircraft.r_pilot_b;
n_pilot_b = n_C_b + ...
    -1/g*(skew(edot(4:6))*(r_pilot_b-rC_b)+skew(omega_b)*skew(omega_b)*(r_pilot_b-rC_b));

[rho,~,~,a] = atmosphere(H);

Mach = V/a;

qbar = 0.5*rho*V^2;

gamma = -2.5;

d_dot = 500*(theta_deg-alpha_deg-gamma);

tau_flare = 2;

if t == 70;
   h_dot = 70/tau_flare;
else
h_dot = -(1/tau_flare)*X(20);
end
Xdot = [edot
    dREOdt
    Phi_dot_rad_s
    Yprop(1)
    X(14)
    X(15)
    X(16)
    X(17)
    X(18)
    d_dot
    h_dot
    X(21)
    X(22)
    X(23)
    X(24)]; 

h_dot = V*sind(theta_deg-alpha_deg);

Y = [V
    alpha_deg
    q_deg_s
    theta_deg
    H
    x
    beta_deg
    phi_deg
    p_deg_s
    r_deg_s
    psi_deg
    y
    n_pilot_b
    n_C_b
    Mach
    qbar
    Yprop(2:end)
    Yaero(4:end)
    Vdot
    alpha_dot_rad_s
    beta_dot_rad_s
    d_dot
    h_dot];

end